stbt-docker

```diff
@@ -0,0 +1,246 @@
+                if best_match is None or len(rel) < len(best_match[1]):
+                    best_match = (hpath, rel)
+        if best_match:
+            return os.path.join(*best_match)
+        else:
+            raise RuntimeError(
+                "Requested container path %s is not in a volume" % path)
+
+
+def find_test_pack_root():
+    root = os.getcwd()
+    while root != '/':
+        if os.path.exists(os.path.join(root, '.stbt.conf')):
+            return root
+        root = os.path.split(root)[0]
+
+
+def dict_to_tarball(d):
+    with tempfile.NamedTemporaryFile() as tarball:
+        with tarfile.open(fileobj=tarball, mode='w') as t:
+            for filename, contents in d.items():
+                ti = tarfile.TarInfo(name=filename)
+                ti.size = len(contents)
+                ti.uid = os.getuid()
+                ti.gid = os.getgid()
+                t.addfile(ti, StringIO(contents))
+        return open(tarball.name, 'r')
+
+DOCKERFILE = """
+FROM {base_image}
+
+RUN mkdir -p /var/lib/stbt/test-pack/config/setup
+ADD setup /var/lib/stbt/test-pack/config/setup/setup
+RUN sudo chown stb-tester:stb-tester /var/lib/stbt/test-pack/config/setup/setup
+RUN chmod a+x /var/lib/stbt/test-pack/config/setup/setup
+RUN /var/lib/stbt/test-pack/config/setup/setup
+"""
+
+
+class DockerNotInstalledError(RuntimeError):
+    pass
+
+
+def build_docker_image(test_pack, stbt_version):
+    base_image = "stbtester/stbt-one-test-pack-base-%s:v24.4" % stbt_version
+    try:
+        with open(os.path.join(test_pack, 'config/setup/setup')) as f:
+            setup = f.read()
+    except IOError as e:
+        if e.errno in [errno.EEXIST, errno.ENOENT]:
+            # No customisation of docker container required
+            return base_image
+        else:
+            raise
+
+    dockerfile = DOCKERFILE.format(base_image=base_image)
+    md5 = hashlib.md5()
+    md5.update(stbt_version)
+    md5.update(setup)
+    docker_image = "stb-tester-test-pack-%s" % md5.hexdigest()
+
+    try:
+        subprocess.check_output(DOCKER + ['inspect', docker_image],
+                                stderr=subprocess.STDOUT)
+        # Customised docker container is already cached
+        return docker_image
+    except subprocess.CalledProcessError:
+        pass
+    except OSError:
+        raise DockerNotInstalledError()
+
+    docker_build_context = {
+        'Dockerfile': dockerfile,
+        'setup': setup,
+    }
+    with dict_to_tarball(docker_build_context) as f:
+        subprocess.check_call(
+            DOCKER + ['build', '-t', docker_image, '-'],
+            stdout=sys.stderr, stdin=f)
+
+    return docker_image
+
+
+def main(argv):
+    root = find_test_pack_root()
+    if root is None:
+        sys.stderr.write(
+            "This command must be run within a test pack.  Couldn't find a "
+            ".stbt.conf in this or any parent directory.\n"
+            "\n"
+            "To create a new test pack run:\n"
```

Comment:
While a page describing the structure of a test-pack would be useful as a reference I wouldn't want to link to that in the instructions on how to get started.  I feel like it would be like linking to [man gitrepository-layout] when all people want to know is `git init`.  I don't particularly want to have instructions to update, although I agree that these are instructions I think they're pretty likely to be stable unless we move away from github.

[man gitrepository-layout]: http://git-scm.com/docs/gitrepository-layout